{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick and clean: Python for biological data processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Day 1, 8:00-16:00, Biological data analysis\n",
    "\n",
    "The first day is also the most essential day in terms of Python programming. It is targeted to custom scientific computing and data mining, so you learn how to adapt a method, or port one from a different language, or glue a remote call to it, also how to find information or mine it using web services. If you are not versed with programming you also get an idea of how to achieve everything with a language.\n",
    "\n",
    "- [What is Python?](day1/what.ipynb)\n",
    "    - How do computers work?\n",
    "    - Python: Stats, strengths and weaknesses\n",
    "    - Python: Past, present and future\n",
    "    - As a new \"pythonista\", who will be your friends?\n",
    "    - Make Python work again!\n",
    "    - Installing libraries\n",
    "    - Python consoles, interpreters and editors\n",
    "    - Python distributions, conda\n",
    "\n",
    "\n",
    "- [Jupyter](day1/jupyter.ipynb)\n",
    "    - History and JupyterLab\n",
    "    - General presentation\n",
    "    - Use in reproducible research\n",
    "    - Other uses\n",
    "    - Slide shows\n",
    "\n",
    "\n",
    "- [The Standard Python tutorial](day1/tutorial.ipynb)\n",
    "    - Basics: Math, Variables, Functions, Control flow, Modules\n",
    "    - Data representation: String, Tuple, List, Set, Dictionary, Objects and Classes\n",
    "    - Standard library modules: script arguments, file operations, timing, processes, forks, multiprocessing\n",
    "\n",
    "\n",
    "- [Python and the data](day1/web.ipynb)\n",
    "    - File IO, streaming, serialization\n",
    "    - Parsing and regular expressions\n",
    "    - Chunking and HDF5, pytables\n",
    "    - XML, HTML editing, web scraping\n",
    "    - Web frameworks.\n",
    "    - SQL frameworks.\n",
    "    - Remote API calls (Entrez, BioBank)\n",
    "\n",
    "\n",
    "- [Python and other languages](day1/languages.ipynb)\n",
    "    - Interacting with other programs through API calls.\n",
    "    - The whole is greater than its parts: syncretism.\n",
    "    - Python and C: make your own API.\n",
    "    - Python and R: direct piping through Jupyter and serialization\n",
    "    - Python and Julia\n",
    "\n",
    "\n",
    "- [Scientific computing](day2/scicomp.ipynb):\n",
    "    - Numpy: advanced array operations\n",
    "    - Scipy introduction: from linear algebra to image analisys\n",
    "    - Simpy: symbolic math\n",
    "    - Networks with networkx: centrality computation\n",
    "    - Fitting a curve, cleaning a signal.\n",
    "\n",
    "\n",
    "- [Statistics](day2/statistics.ipynb):\n",
    "    - Dataframing with pandas\n",
    "    - Normalizing a dataset\n",
    "    - Back to Scipy: ANOVA\n",
    "    - Statistical tests: Enrichment Analysis\n",
    "    - Statsmodels: Differential expression\n",
    "\n",
    "\n",
    "- [Visualization](day2/visualization.ipynb):\n",
    "    - Standard plots with matplotlib and seaborn: line, scatter, chart\n",
    "    - Web publishing with plotly and bokeh: heatmap example\n",
    "    - Network display with networkx\n",
    "    - GUI programming with wxpython\n",
    "    - Web interfaces\n",
    "\n",
    "\n",
    "- [Optimization]()\n",
    "    - Least squares\n",
    "    - Gradient descent\n",
    "    - Constraint optimization\n",
    "    - Integer programming\n",
    "\n",
    "\n",
    "- [Dynamical systems]()\n",
    "    - Time series\n",
    "    - Fixed points and bifurcations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Day 2, 8:00-16:00, Biological data science\n",
    "\n",
    "This day has a smaller focus on actual programing and a more practical focus on how to perform machine learning, statistical learning and patern recognition. This day builds on the \"science stack\" libraries and makes heavy use of scikit-learn and other more exotic libraries.\n",
    "\n",
    "- [Intro to data science](day2/introds.ipynb)\n",
    "    - Data science vs biology.\n",
    "    - How to extract information from data.\n",
    "    - Dataset, model, prediction.\n",
    "    - Supervised vs Unsupervised ML\n",
    "    - Linearity, Nonlinearity\n",
    "    - Data distributions in biology.\n",
    "\n",
    "\n",
    "- [Dimension Reduction]():\n",
    "    - linear: SVD, PCA, ICA, FA\n",
    "    - nonlinear: t-SNE\n",
    "\n",
    "\n",
    "- [Clustering]():\n",
    "    - Similarity scores in various biological data\n",
    "    - Useful methods: Kmeans, hierarchical, spectral, DBSCAN\n",
    "    - Graph community detection: modularity, random walking, infomap\n",
    "\n",
    "\n",
    "- [Regression]():\n",
    "    - models: OLS, Lasso, Ridge, Bayesian, etc\n",
    "    - Model validation, Hyperparametrization.\n",
    "\n",
    "\n",
    "- [Classification]():\n",
    "    - Logistic regression\n",
    "    - SVD, Kernel methods\n",
    "    - PLS, OPLS, OPLS-DA\n",
    "    - Decision trees and random forests\n",
    "\n",
    "\n",
    "- [Data integration]():\n",
    "    - Integrative clustering\n",
    "    - Covariance optimization\n",
    "    - Bayesian learning\n",
    "\n",
    "\n",
    "- [Deep learning]()\n",
    "    - Essentials, Tensor math\n",
    "    - One example: Transcriptional regulation\n",
    "    - Keras, Tensorflow and PyTorch\n",
    "\n",
    "\n",
    "- [Statistical learning]()\n",
    "    - Bayesian statistics\n",
    "    - PyMC3 basics\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 3, 8:00-16:00, Biological data engineering\n",
    "\n",
    "We covered the more advanced aspects of Python, and explored some of the libraries that make data science work. This day is dedicated to engineering the computing infrastructure and Python's role in it. What is the state of the computing infrastructure today, how to use Python to organize your work with reproducibility in mind, how to run it on clouds and GPU machines?\n",
    "\n",
    "- [Gentle introduction]\n",
    "    - PC, Server, Grid, Cloud, IoT\n",
    "    - Desktop OS and Python.\n",
    "    - Using Docker.\n",
    "    - Using Jupyter and Python inside Docker.\n",
    "\n",
    "\n",
    "- [DevOps](day1/devops.ipynb)\n",
    "    - Reproducible research.\n",
    "    - Continuous integration.\n",
    "    - How to source your code.\n",
    "    - Using source editors. What matters?\n",
    "    - Distributed version control using git.\n",
    "    - Development vs production.\n",
    "    - When do we need containers? Using Docker.\n",
    "    - Speed: Profiling, IPython, JIT \n",
    "    - Robustness: unit testing\n",
    "    - Documentation: pydoc and Sphinx\n",
    "\n",
    "\n",
    "- [Workflow management](day2/workflow.ipynb)\n",
    "    - Snakemake tutorial\n",
    "    - Other examples: nextflow, luigi.\n",
    "    - Make your own workflow in Python.\n",
    "\n",
    "\n",
    "- [Grid computing]()\n",
    "    - What is HPC?\n",
    "    - Singularity.\n",
    "\n",
    "\n",
    "- [Distributed computing](day1/cloud.ipynb)\n",
    "    - What is the cloud?\n",
    "    - Hadoop and Spark\n",
    "    - Spinning instances: Back to containers.\n",
    "    - Serverless.\n",
    "\n",
    "\n",
    "- [AWS cloud]()\n",
    "    - Setting up a reserved instance.\n",
    "    - S3 Buckets.\n",
    "    - Access spot instance.\n",
    "    - Spin-up containers.\n",
    "\n",
    "\n",
    "- [Spark]()\n",
    "    - MapReduce, Hadoop and Spark.\n",
    "    - Data lake: Hive, Pigs, YARN.\n",
    "    - Setup a Spark cluster in AWS.\n",
    "    - ML with PySpark\n",
    "\n",
    "\n",
    "- [GPU acceleration on the cloud]()\n",
    "    - GPU computing, FPGA, ASICS\n",
    "    - Statistical learning example: PyMC3\n",
    "    - Deep Learning example: Keras, Tensorflow\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task day, 8:00-16:00, 'Omics\n",
    "\n",
    "We setup the problems and describe the tasks, and give you some helper code to start with, and you will work on your picked task in class. I will tend to guide rather than tell. You are engouraged to bring your own task, but if you want the quality of advice to be high it would be good to send me a description or contact me in advance! We will also have an individual follow up at some later date.\n",
    "\n",
    "- [Sequencing - NGS pipelining](day3/ngs_pipelining.ipynb):\n",
    "    - Open a cloud instance and install required programs\n",
    "    - Setup the pipeline\n",
    "    - Read mapping and IGV inspection\n",
    "    - Normalizing counts and differential expression\n",
    "    - Galaxy integration\n",
    "    \n",
    "    \n",
    "- [Sequencing - Biopython](day3/sequencing.ipynb)\n",
    "    - Make a toy sequencing library in standard Python for processing DNA, RNA and protein data.\n",
    "    - Implement the DNA, RNA and proteins as Python classes\n",
    "    - Make methods for transcription, translation, regulation.\n",
    "    - Compute several sequence similarity scores, such as hamming distance and mutual information.\n",
    "    - Add BioPython methods and prefix them with bp\n",
    "    - Describe your module in a tutorial like fashion\n",
    "    \n",
    "    \n",
    "- [Gene Expression](day3/expression.ipynb)\n",
    "    - Download a GEO dataset and prepare it\n",
    "    - Cluster the genes based on their expression\n",
    "    - Compute a co-expression network\n",
    "    - Compute differential gene expression for a set of samples.\n",
    "    - Compute functional enrichment of the main clusters.\n",
    "    \n",
    "    \n",
    "- [Transcriptomics](day3/transcriptomics.ipynb)\n",
    "    - Extract the promoter regions using biopython\n",
    "    - Investigate de-novo motifs on clusters of genes using meme or steme\n",
    "    - Use TransFac database to search for motif occurences on selected genes\n",
    "    - Save the found motifs and related data\n",
    "    - Test which of the motif occurences on your selected genes is significant\n",
    "    \n",
    "    \n",
    "- [Proteomics](day3/proteomics.ipynb)\n",
    "    - Compute a protein similarity graph, cluster enrichment study\n",
    "    - Perform structural alignment and plots with PyMol\n",
    "    \n",
    "    \n",
    "- [Metabolomics](day3/metabolomics.ipynb)\n",
    "    - Metabolic pathway assembly, enrichment and display\n",
    "    - Flux balance analysis\n",
    "    \n",
    "    \n",
    "- [Dynamic modeling](day3/dynamic_modeling.ipynb)\n",
    "    - Load a curated SBML model\n",
    "    - Plot the model\n",
    "    - Solve the model\n",
    "    - Peak identification\n",
    "    - Pathway studies\n",
    "    \n",
    "    \n",
    "- [Population genetics and philogeny](PopGen.ipynb) - Removed/Not included.\n",
    "    - Run a small scale coalescent simulation\n",
    "    - Compute a philogeny tree and display it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
